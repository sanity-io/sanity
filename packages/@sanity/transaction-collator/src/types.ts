export type Mutation = {
  create?: {_id: string}
  createIfNotExists?: {_id: string}
  createOrReplace?: {_id: string}
  createSquashed?: {_id: string, authors: string[], document: {_id: string}}
  delete?: {id: string}
  patch?: {id: string}
}

export type EventType = 'created' | 'edited' | 'published' | 'unpublished' | 'discardDraft' | 'truncated' | 'deleted' | 'restored' | 'unknown'

export type HistoryEvent = {
  displayDocumentId: string | null
  documentIDs: string[]
  endTime: string,
  rev: string
  startTime: string
  transactionIds: string[]
  type: EventType
  userIds: string[]
}

export type Transaction = {
  author: string
  documentIDs: string[]
  id: string
  mutations: Mutation[]
  timestamp: string
}
