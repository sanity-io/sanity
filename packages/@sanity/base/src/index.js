import DefaultRootComponent from './components/DefaultRootComponent'

export {DefaultRootComponent}
