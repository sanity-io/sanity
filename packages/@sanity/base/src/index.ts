export {default as DefaultRootComponent} from './components/DefaultRootComponent'
export {UserStore} from './datastores/user/types'
