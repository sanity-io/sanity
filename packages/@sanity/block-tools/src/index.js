// @flow

import blockContentTypeToOptions from './util/blockContentTypeToOptions'
import blocksToSlateState from './converters/blocksToSlateState'
import HtmlDeserializer from './HtmlDeserializer'
import slateStateToBlocks from './converters/slateStateToBlocks'

/**
 * BlockTools - various tools for Sanity block content
 *
 * @param {Object} blockContentType
 *    The compiled schema for the block content type to work with
 *
 */
export default {
  /**
   * Convert HTML to blocks respecting the block content type's schema
   *
   * @param {String} html
   *
   * @param {Object} options
   *   @property {Object} blockContentType
   *      A compiled version of the schema type for the block content
   *   @property {Array} rules
   *      Optional rules working on the HTML (will be ruled first)
   *   @property {Function} parseHtml
   *      API compatible model as returned from DOMParser for using server side.
   * @returns {Array} Blocks
   */
  htmlToBlocks(html, options = {}) {
    const deserializer = new HtmlDeserializer(options)
    return deserializer.deserialize(html)
  },

  /**
   * Convert a serialized Slate state to blocks
   *
   * @param {Object} An object representing the structure of the Slate JSON.
   * @param {Object} blockContentType
   * @returns {Array} Blocks
   */
  slateStateToBlocks(slateJson, blockContentType) {
    return slateStateToBlocks(slateJson, blockContentType)
  },

  /**
   * Convert blocks to a serialized Slate state
   *
   * @param {Array} blocks
   * @param {Object} blockContentType
   * @returns {Object} An object representing the serialized Slate state.
   */

  blocksToSlateState(blocks, blockContentType) {
    return blocksToSlateState(blocks, blockContentType)
  },

  /**
   * Returns the feature-set of a compiled block content type.
   *
   * @param {Object} blockContentType
   * @returns {Object} The feature-set
   */
  getBlockContentFeatures(blockType) {
    return blockContentTypeToOptions(blockType)
  }
}
