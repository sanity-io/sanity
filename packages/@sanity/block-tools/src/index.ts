import type {ArraySchemaType, PortableTextTextBlock} from '@sanity/types'
import blockContentTypeFeatures from './util/blockContentTypeFeatures'
import HtmlDeserializer from './HtmlDeserializer'
import {normalizeBlock} from './util/normalizeBlock'
import {BlockContentFeatures, HtmlDeserializerOptions, TypedObject} from './types'

/**
 * Convert HTML to blocks respecting the block content type's schema
 *
 * @param html - The HTML to convert to blocks
 * @param blockContentType - A compiled version of the schema type for the block content
 * @param options - Options for deserializing HTML to blocks
 * @returns Array of blocks
 * @public
 */
export function htmlToBlocks(
  html: string,
  blockContentType: ArraySchemaType,
  options: HtmlDeserializerOptions = {},
): (TypedObject | PortableTextTextBlock)[] {
  const deserializer = new HtmlDeserializer(blockContentType, options)
  return deserializer.deserialize(html).map((block) => normalizeBlock(block))
}

/**
 * Normalize and extract features of an schema type containing a block type
 *
 * @param blockContentType - Schema type for the block type
 * @returns Returns the featureset of a compiled block content type.
 * @public
 */
export function getBlockContentFeatures(blockContentType: ArraySchemaType): BlockContentFeatures {
  return blockContentTypeFeatures(blockContentType)
}

export {normalizeBlock}
export {randomKey} from './util/randomKey'
export type {TypedObject, HtmlDeserializerOptions, BlockContentFeatures}
export type {
  ResolvedAnnotationType,
  DeserializerRule,
  HtmlParser,
  ArbitraryTypedObject,
  BlockEditorSchemaProps,
} from './types'
export type {BlockNormalizationOptions} from './util/normalizeBlock'
