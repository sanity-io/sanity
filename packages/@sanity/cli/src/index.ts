export * from './types'
export * from './config'

export type {CliClientOptions} from './cliClient'
export {getCliClient} from './cliClient'
export {loadEnv} from './util/loadEnv'

export type {SanityClient} from '@sanity/client'
