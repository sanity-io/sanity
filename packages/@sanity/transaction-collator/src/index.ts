export type {HistoryEvent} from './types'
export {transactionsToEvents} from './transactionsToEvents'
