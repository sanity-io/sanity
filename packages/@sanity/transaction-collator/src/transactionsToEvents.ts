import {isEqual, uniq} from 'lodash'
import {EventType, HistoryEvent, Transaction, Mutation} from './types'
import {ndjsonToArray} from './utils/ndjsonToArray'

const EDIT_EVENT_TIME_TRESHHOLD_MS = 1000 * 60 * 5 // 5 minutes

export function transactionsToEvents(
  documentIds: string[],
  transactions: string | Buffer | Transaction[]
): HistoryEvent[] {
  const rawItems = Array.isArray(transactions) ? transactions : ndjsonToArray(transactions)
  return (
    rawItems
      // Make sure we only deal with transactions that are relevant for our documents
      .filter(
        (transaction: Transaction) =>
          transaction.documentIDs && transaction.documentIDs.some((id) => documentIds.includes(id))
      )
      // ensure transactions are sorted by time
      .sort(compareTimestamp)
      // Turn a transaction into a classified HistoryEvent
      .map((transaction, index) => mapToEvents(transaction, documentIds, index))
      // Chunk and group edit events
      .reduce(reduceEdits, [])
      // Manipulate truncation events to be able to restore to published version
      .reduce(createReduceTruncatedFn(), [])
  )
}

function mapToEvents(transaction: Transaction, documentIds: string[], index = 0): HistoryEvent {
  const {type, documentId} = mutationsToEventTypeAndDocumentId(
    filterRelevantMutations(transaction.mutations, documentIds),
    index
  )
  const timestamp = transaction.timestamp
  const userIds = findUserIds(transaction, type)
  return {
    type,
    documentIDs: transaction.documentIDs,
    displayDocumentId: documentId,
    rev: transaction.id,
    userIds,
    transactionIds: [transaction.id],
    startTime: timestamp,
    endTime: timestamp,
  }
}

function reduceEdits(
  acc: HistoryEvent[],
  current: HistoryEvent,
  index: number,
  arr: HistoryEvent[]
) {
  const nextEvent = arr[index + 1]
  const skipEvent =
    current.type === 'edited' &&
    nextEvent &&
    nextEvent.type === 'edited' &&
    new Date(nextEvent.endTime).getTime() - new Date(current.endTime).getTime() <
      EDIT_EVENT_TIME_TRESHHOLD_MS &&
    isEqual(current.documentIDs, nextEvent.documentIDs)
  if (skipEvent) {
    // Lift authors over to next event
    nextEvent.userIds = uniq(nextEvent.userIds.concat(current.userIds))
    // Lift list of transactions over to next event
    nextEvent.transactionIds = uniq(current.transactionIds.concat(nextEvent.transactionIds))
    // Set startTime on next event to be this one if not done already
    // (then startTime and endTime would be different)
    if (current.startTime === current.endTime) {
      nextEvent.startTime = current.startTime
    }
  } else {
    acc.push(current)
  }
  return acc
}

function createReduceTruncatedFn() {
  let truncated: HistoryEvent[] | undefined
  return (acc: HistoryEvent[], current: HistoryEvent, index: number, arr: HistoryEvent[]) => {
    truncated = truncated || arr.filter((event) => event.type === 'truncated')
    if (!truncated.includes(current)) {
      acc.push(current)
    }
    if (index === arr.length - 1) {
      const draftTruncationEvent = truncated.find(
        (evt) => !!evt.displayDocumentId && evt.displayDocumentId.startsWith('drafts.')
      )
      const publishedTruncationEvent = truncated.find(
        (evt) => !!evt.displayDocumentId && !evt.displayDocumentId.startsWith('drafts.')
      )
      if (draftTruncationEvent && publishedTruncationEvent) {
        acc.unshift({...draftTruncationEvent, type: 'edited'})
        acc.unshift(publishedTruncationEvent)
      } else if (publishedTruncationEvent) {
        acc.unshift(publishedTruncationEvent)
      } else if (draftTruncationEvent) {
        acc.unshift(draftTruncationEvent)
      }
    }
    return acc
  }
}

export function mutationsToEventTypeAndDocumentId(
  mutations: Mutation[],
  transactionIndex: number
): {type: EventType; documentId: string | null} {
  const withoutPatches = mutations.filter((mut) => mut.patch === undefined)
  const createOrReplaceMutation = withoutPatches.find((mut) => mut.createOrReplace !== undefined)
  const createOrReplacePatch = createOrReplaceMutation && createOrReplaceMutation.createOrReplace

  const createMutation = withoutPatches.find((mut) => mut.create !== undefined)
  const createPatch = createMutation && createMutation.create

  const createIfNotExistsMutation = withoutPatches.find(
    (mut) => mut.createIfNotExists !== undefined
  )
  const createIfNotExistsPatch =
    createIfNotExistsMutation && createIfNotExistsMutation.createIfNotExists

  const deleteMutation = withoutPatches.find((mut) => mut.delete !== undefined)
  const deletePatch = deleteMutation && deleteMutation.delete

  const squashedMutation = withoutPatches.find((mut) => mut.createSquashed !== undefined)
  const squashedPatch = squashedMutation && squashedMutation.createSquashed

  const createValue = createOrReplacePatch || createPatch || createIfNotExistsPatch

  // Created
  if (transactionIndex === 0) {
    const type = 'created'
    if (createOrReplacePatch) {
      return {type, documentId: createOrReplacePatch._id}
    }
    if (createIfNotExistsPatch) {
      return {type, documentId: createIfNotExistsPatch._id}
    }
    if (createPatch) {
      return {type, documentId: createPatch._id}
    }
  }

  // (re) created
  if (transactionIndex > 0 && mutations.length === 1 && createIfNotExistsPatch) {
    const type = createIfNotExistsPatch._id.startsWith('.draft') ? 'edited' : 'published'
    return {type, documentId: createIfNotExistsPatch._id}
  }

  // Published
  if (
    (createOrReplacePatch || createPatch || createIfNotExistsPatch) &&
    deletePatch &&
    deletePatch.id.startsWith('drafts.')
  ) {
    return {
      type: 'published',
      documentId: (createValue && createValue._id) || null,
    }
  }

  // Unpublished
  if (
    withoutPatches.length === 2 &&
    (createIfNotExistsPatch || createPatch) &&
    deletePatch &&
    !deletePatch.id.startsWith('drafts.')
  ) {
    return {
      type: 'unpublished',
      documentId: (createValue && createValue._id) || null,
    }
  }

  // Restored to previous version
  if (
    (createOrReplacePatch && createOrReplacePatch._id.startsWith('drafts.')) ||
    (createPatch && createPatch._id.startsWith('drafts.')) ||
    (createIfNotExistsPatch && createIfNotExistsPatch._id.startsWith('drafts.'))
  ) {
    return {
      type: 'edited',
      documentId: (createValue && createValue._id) || null,
    }
  }

  // Discard drafted changes
  if (mutations.length === 1 && deletePatch && deletePatch.id.startsWith('drafts.')) {
    return {type: 'discardDraft', documentId: deletePatch.id.replace('drafts.', '')}
  }

  // Truncated history
  if (mutations.length === 1 && squashedPatch) {
    return {type: 'truncated', documentId: squashedPatch.document._id}
  }

  // Deleted
  if (mutations.every((mut) => mut.delete !== undefined)) {
    return {type: 'deleted', documentId: null}
  }

  // Edited
  const patchedMutation = mutations.find((mut) => mut.patch !== undefined)
  if (patchedMutation && patchedMutation.patch) {
    return {type: 'edited', documentId: patchedMutation.patch.id}
  }

  // Edited (createOrReplace)
  if (createOrReplacePatch) {
    return {type: 'edited', documentId: createOrReplacePatch._id}
  }

  return {type: 'unknown', documentId: null}
}

function findUserIds(transaction: Transaction, type: EventType): string[] {
  // The truncated event is kind of special
  if (type === 'truncated') {
    const createSquasedMut = transaction.mutations.find((mut) => mut.createSquashed !== undefined)
    const createSquasedPatch = createSquasedMut && createSquasedMut.createSquashed
    if (createSquasedPatch) {
      return createSquasedPatch.authors
    }
  }
  // Default is to return the transaction author
  return [transaction.author]
}

function compareTimestamp(a: Transaction, b: Transaction) {
  return new Date(a.timestamp).getTime() - new Date(b.timestamp).getTime()
}

function filterRelevantMutations(mutations: Mutation[], documentIds: string[]) {
  return mutations.filter((mut) => {
    return Object.keys(mut)
      .map((key) => {
        const val = (mut as any)[key]
        return val.id || val._id || (val.document && val.document._id) || false
      })
      .some((id) => id && documentIds.includes(id))
  })
}
