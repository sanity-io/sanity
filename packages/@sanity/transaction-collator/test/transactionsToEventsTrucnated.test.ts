import {readFileSync} from 'fs'
import {transactionsToEvents} from '../src/index'
const ndJSON = readFileSync('./test/fixtures/transactions_truncated.ndjson', 'utf8')
const documentIds = [
  'team-f0b2a4cf7a936ac478f8ec25d341d1edbd592ab5',
  'drafts.team-f0b2a4cf7a936ac478f8ec25d341d1edbd592ab5',
]

test('classifies transactions to events ', () => {
  const result = transactionsToEvents(documentIds, ndJSON)
  expect(result).toMatchSnapshot()
})
