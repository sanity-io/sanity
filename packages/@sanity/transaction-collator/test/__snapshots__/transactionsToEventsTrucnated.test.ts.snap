// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`classifies transactions to events  1`] = `
Array [
  Object {
    "displayDocumentId": "team-f0b2a4cf7a936ac478f8ec25d341d1edbd592ab5",
    "documentIDs": Array [
      "team-42ee9c1ca3d38337a3c68c296b4c86b863a2eaf3",
      "team-8673bdc6e014b850ae030e123aa058532271e13e",
      "team-8eb00df53fb2e43bd19bd7ae09f3e89e5055b8a5",
      "team-07e85992ace967c0d3b5bb30450781e1d45e313d",
      "team-7f79e928590566c20a31627936e5f94001f20a07",
      "team-723c34e02fabd0ce26d65908030f03c5eb3a02ea",
      "team-32198203519a4c1878d27d6f8021ed310ea88548",
      "team-83cb86ed745860955d4543de578cef04565aec2f",
      "team-25cfd85e6e70a7d34920dcb2e460d4e0ea0a189d",
      "team-64ec68d3bae4d32db7a3f4b074f7f62ca6280341",
      "team-a97bccb11bd711835301448975228bf6a400b504",
      "team-6b2720d83daa72a930d58fa3743eb794010590cc",
      "team-d95b7af9cc70571269d5f7a852b7c14593e9b1df",
      "team-eadd76229b8bc763cab2d4915fe5dc2b8f997590",
      "team-673b6c5a941c04243169ce40de69aef331e61892",
      "team-68a12e49390d1e2084cf22eda67291d666528f7e",
      "team-be5bee0fce3fd5a0b069d5901fd36b7750cfbd20",
      "team-c11aacd85858758c2aae5a00be1713d98fc8dda3",
      "team-00c8a94ab93f5fa977311dc04897ddcfed791193",
      "team-568be2771a60a817365c290619ee9245875b919e",
      "team-cff50ae7358ba9938faf05896f5fc4ada4fd06e4",
      "team-2f4bd42e4112072310acf63a1edf011388ee42a9",
      "team-beda50c6fc6e9ed24996e21b98877a1071e5f6ec",
      "team-f0b2a4cf7a936ac478f8ec25d341d1edbd592ab5",
      "team-8fe8681997c70fa79d6402d73b39d70cf6a02ec3",
      "team-f45f06759142188d7ffa854a5c4556035e81e7dc",
      "team-16f40260ed45426dcd43664a66395ecd50b6b9e5",
      "team-caa8274de32a4cd6f6c1082c5790021888fe350e",
    ],
    "endTime": "2018-03-16T17:47:15.108594Z",
    "rev": "nlwJW0GjnaISO6pLcPcRB2",
    "startTime": "2018-03-16T17:47:15.108594Z",
    "transactionIds": Array [
      "nlwJW0GjnaISO6pLcPcRB2",
    ],
    "type": "truncated",
    "userIds": Array [
      "pEdMA3tnN",
    ],
  },
  Object {
    "displayDocumentId": "drafts.team-f0b2a4cf7a936ac478f8ec25d341d1edbd592ab5",
    "documentIDs": Array [
      "drafts.team-f0b2a4cf7a936ac478f8ec25d341d1edbd592ab5",
    ],
    "endTime": "2018-03-16T18:43:19.078619Z",
    "rev": "g22lfg-et7-5a5-sje-eayjevqpn",
    "startTime": "2018-03-16T18:43:19.078619Z",
    "transactionIds": Array [
      "g22lfg-et7-5a5-sje-eayjevqpn",
    ],
    "type": "edited",
    "userIds": Array [
      "pEdMA3tnN",
    ],
  },
  Object {
    "displayDocumentId": "drafts.team-f0b2a4cf7a936ac478f8ec25d341d1edbd592ab5",
    "documentIDs": Array [
      "drafts.team-f0b2a4cf7a936ac478f8ec25d341d1edbd592ab5",
    ],
    "endTime": "2019-06-27T12:55:33.470807Z",
    "rev": "vkjpts-q8n-08d-g88-ez4kmnwg4",
    "startTime": "2019-06-27T12:55:32.469439Z",
    "transactionIds": Array [
      "foa4ka-n7k-zqm-4j3-3u9cpzgi9",
      "vttnol-c7p-9qu-o0j-jvyldf0fc",
      "4fzqow-l9h-qt6-96r-nh3pk0cf7",
      "vkjpts-q8n-08d-g88-ez4kmnwg4",
    ],
    "type": "edited",
    "userIds": Array [
      "pesMTOkCm",
    ],
  },
]
`;
